### ISSUE TITLE
Restore-DbaDatabase fails if source DB was in SIMPLE RecoveryMode and you specify a RecoveryTime

### ISSUE THREAD

**Issue by ashdar** [sqlcollaborative/dbatools#3469]
### General Troubleshooting steps
Tested in 0.9.318

- [X] Verified running the latest release of dbatools?
- [X] Verified errors are not related to permissions?
- [X] Can duplicate in new/clean PowerShell session (clean = `powershell -NoProfile`)?

### Version Information
 - Operating System (Windows 10|Windows7): 
 - PowerShell Version: 5.1.14409.1012 
 - SQL Server (Developer|10.50.6220): 

## Steps to Reproduce

In a situation where you are restoring a FULL backup plus one out of several DIFF backups, Restore-DbaDatabase fails if the source DB was in SIMPLE RecoveryMode and you specify a RecoveryTime in order to filter out DIFF backups that are "too new". 

To sum up: The restore command fails with a syntax error because dbatools includes STOPAT in the RESTORE database command, even though the DB is in simple mode.

**Here is the scenario:**
1. Database ABC is running in SIMPLE mode.
2. We are using Hallengren-style backups
3. A full backup occurs Sunday mornings at 6 AM. Let's call that 2/4/2018 06:00.
4. A DIFF backup occurs every other day, Monday through Saturday, at 6 AM. We have backups from 2/5/2018 through 2/10/2018, always at 6 AM.
5. (Pretend that) today is 2/13/2018. We want to restore the ABC database to a DEV server, but only up to Wednesday (2/7/2018) morning
6. Restore-DbaDatabase -SqlInstance TestSrv -database ABC -MaintenanceSolution -path \\backupsvr\backups -RestoreTime '2/7/2018 9:00 AM'
7. This command finds the FULL backup from 2/4. **Good**. This command also finds the most recent DIFF backup that is older than our RestoreTime. That would be the DIFF from 2/7/2018 06:00. **Good**.
9. **Problem**: The RESTORE command for the full backup file is incorrect because it includes a STOPAT value. You can easily see this by using -OutputScriptOnly. STOPAT is not legal syntax for a restore of a full backup file for a database running in simple recovery mode. STOPAT is legal syntax for a restore of a full backup of a database running in bulk logged or fully logged recovery modes. 

**My proposed fix:**
1. BackupHistory.cs needs to be enhanced to add the RecoveryModel attribute to the Sqlcollaborative.Dbatools.Database.BackupHistory class.
2. Get-DbaBackupInformation needs to be enhanced to output the RecoveryModel in addition to the attributes it already provides
3. Invoke-DbaAdvancedRestore needs to check if backup type (which is already available) is full and RecoveryModel (which requires the first change) is simple, if so then do not set a ToPointInTime on the $restore object. 

### Additional information
I've got code changed and ready to go in my repo, I am looking for someone to reassure me that this is a sane thing to do and I'll submit a PR.

**Comment by niphlod** [comment]
@ashdar: thanks for the report.

@Stuart-Moore will probably comment better (and I'm going to bed), but just my 2 cents..... 

Can't we shortcircuit typechecking the history (which is held in `recovery_model` within msdb, while `RecoveryModel` when reading headers) not setting `$restore.ToPointInTime` when we do not have a log backup in the pipe ?

tl;dr: if the logic for selecting the correct set of files is correct and done before invoking the actual restore, isn't it easier to add the skip in PS rather than "formally correcting" the whole pipeline of commands ? Because when you want to have it formal, we'd probably have to account also for `has_bulk_logged_data/HasBulkLoggedData` .... etc .

**Comment by ashdar** [comment]
I am not familiar enough with the restore code to make much sense; I learned only as much as I needed to to fix the problem that I saw at one of my clients. I've actually been sitting on this for a few weeks because I haven't had time to actually work on it. I did test the code with two scenarios:

1. Restore a DB in simple by providing a 'RecoverTime' value that finds just one full backup file and ignores two subsequent differential backups.
2. Restore a DB in simple by providing a 'RecoverTime' value that finds one full backup file and one differential backup, but ignores the subsequent differential backup.

I hadn't thought about bulk-logged, but I've just double-checked restoration of a bulk logged database. The following command does not cause a syntax error, like it would with a DB in simple recovery mode, so I think that we are good just looking for SIMPLE recovery mode in the 'Advanced Restore' cmdlet.

`
restore database bulklogged_test
from disk = 'c:\temp\backups\bulklogged_test\full\bulklogged_test_full_20180404.bak'
with stopat = '2018-04-04 09:17:00'
`

**Comment by Stuart-Moore** [comment]
@niphlod - that's my thought as well.

Dropping something in Invoke-DbaAdvancedRestore between L172 and L195 would
be the best way. Just working out the simplest way to shove it in there.

**Comment by Stuart-Moore** [comment]
Just pushed https://github.com/sqlcollaborative/dbatools/tree/restdiffsimple
which
has a fix that works for my small test rig.

If you could test and confirm it works i'll push up a pester test as well
to stop it getting broken again.

Cheers
Stuart

**Comment by ashdar** [comment]
My change takes L182 to L184 from:
`if ($restoretime -gt (Get-Date) -or $Restore.RestoreTime -gt (Get-Date)) {
$Restore.ToPointInTime = $null	                     $Restore.ToPointInTime = $null
}`

To:
`if ($restoretime -gt (Get-Date) -or $Restore.RestoreTime -gt (Get-Date) -or ($backup.Type -eq 'Database' -and $backup.RecoveryModel -eq 'Simple')) {
$Restore.ToPointInTime = $null
}`

**Comment by ashdar** [comment]
I got the restdiffsimple branch onto my workstation. 

If I run this on a PowerShell command line:
`Get-DbaBackupInformation -SqlInstance $script:instance1 -Path c:\temp\backups\simple_test | sort End`

I get this:
![capture1](https://user-images.githubusercontent.com/11427665/38314315-e1c52618-37f3-11e8-8008-08af35a62e80.PNG)

I want to restore to where the database was at 9:27 AM, which is between when the two differentials ran, so I run this on a PowerShell command line:
`Restore-DbaDatabase -MaintenanceSolution -path c:\temp\backups\simple_test -server
$script:instance1 -DestinationFilePrefix 'Restore-DbaDatabase' -database 'simple_test'  -RestoreTime '2018-04-04 09:27:00'`

I believe that should restore the full backup from 9:18:40 AM and the diff backup from 9:24:33 AM. Instead, I get a failure:
![capture2](https://user-images.githubusercontent.com/11427665/38314601-9b07df4e-37f4-11e8-94cb-dcb0fd2324f9.PNG)


1. Looking at the script output,  I expect there to be a RESTORE for the full and a RESTORE for the DIFF. There is only a RESTORE for the full. 
2. The RESTORE for the full has a STOPAT value, which is a runtime error because STOPAT is not compatible with restoring databases that were simple recovery mode when they were backed up. 
3. There is an exception, and the restore fails. 99% this is because of the STOPAT.

I'm not very familiar with GitHub. [My proposed change is sitting in a branch on my fork](https://github.com/ashdar/dbatools/compare/master...ashdar:bugfix-Restore-DbaDatabase). Can you see it?

**Comment by Stuart-Moore** [comment]
Morning,
There's a guide to submitting you're first PR here - https://github.com/sqlcollaborative/dbatools/wiki/Your-First-Pull-Request

If you follow that through then it'll end up in the approval queue once it's been tested against our test rig

**Comment by niphlod** [comment]
I pushed a fix to the https://github.com/sqlcollaborative/dbatools/tree/restdiffsimple tree, take a look at that. Basically it doesn't bother setting topointintime if there is no incremental in the chain

**Comment by ashdar** [comment]
@Stuart-Moore, I don't see anything about how to deal with dlls. I've modified the backuphistory.cs file, which requires rebuilding the dbatools.dll. Do I include updated copies of the .\bin and .\debug dlls, or is that built somehow by devops magic? This is my second PR ever and it's kind of intimidating.

**Comment by ashdar** [comment]
@niphlod. I will look at that now.

**Comment by ashdar** [comment]
After pulling down the latest restdiffsimple, I'm not seeing a change 
![capture3](https://user-images.githubusercontent.com/11427665/38388216-9469e7c8-38e8-11e8-89ba-2003936ee489.PNG)

**Comment by niphlod** [comment]
hum, something is definitely going on. any chance we can chat on slack ?

**Comment by ashdar** [comment]
Happy to give it a try, but I'm on the US East coast, 6 hours behind Milan and I'm guessing we will have to try tomorrow. I don't normally hang out on slack, but I can get to #dbatools channel OK.

**Comment by niphlod** [comment]
no probls

**Comment by Stuart-Moore** [comment]
restdiffsimple doesn't contain any dll changes, should it?

There's a nack to altering the version number to get it recompiled.

**Comment by Stuart-Moore** [comment]
PR in. @ashdar afraid I couldn't tag you as reviewer but if you could try it that'd be great. Though I've included a new appv test that covers it.

@niphlod 
To force the dll to recompile you have to remove it, and then update the version numbers in library.ps1 and AssemblyInfo.cs. Looks like fred dropped some docu into library.ps1 since I last looked.

**Comment by ashdar** [comment]
If I had realized what a PITA changing the DLL is, I would not have even tried. It seems easy enough, just add a string attribute to BackupHistory.cs, right?

I've been overwhelmed all morning with pop-up tasks and the first Friday of a month is always busy for me, I'll try to have a look in the next four hours.

**Comment by niphlod** [comment]
No probl. Most of times is easy peasy but if you went and raised a PR with
"errors" we'd have swooped and fixed step by step throughout the review
process.
Thanks for the contribution ;-)

Il ven 6 apr 2018, 18:50 dstrait <notifications@github.com> ha scritto:

**Comment by ashdar** [comment]
That passes my tests, which aren't dependent on any appveyor configuration. 

![image](https://user-images.githubusercontent.com/11427665/38436988-1db3634a-39a4-11e8-893f-ef408759cc53.png)

**Comment by Zhunya** [comment]
@ashdar be so kind, share your test code.

**Comment by ashdar** [comment]
Will this [gist ](https://gist.github.com/ashdar/a5c0e6cf34f56daf310381fc0dae975a) work for you?

**Comment by Zhunya** [comment]
Brilliant, thank you!
